src/application/controller/controller.c

```diff
@@ -1,2 +1,118 @@
 #include "application/controller/controller.h"
 
+#include "application/flight_phase/flight_phase.h"
+#include "application/logger/log.h"
+#include "queue.h"
+
+static QueueHandle_t internal_state_queue;
+static QueueHandle_t output_queue;
+
+#define CONTROLLER_CYCLE_TIMEOUT_MS 5
+
+static controller_t controller_state = {0};
+static controller_input_t controller_input __attribute__((unused)) = {0};
+static controller_output_t controller_output = {0};
+static controller_gain_t controller_gain __attribute__((unused)) = {0};
+
+/*
+    TODO Send `canard_angle`, the desired canard angle (radians) to CAN
+*/
+static w_status_t controller_send_can(float canard_angle) {
+    // Build the CAN msg using [canard-specific canlib function to be defined later].
+
+    // Send this to can handler module’s tx
+    (void)canard_angle;
+    return W_SUCCESS;
+}
+
+/**
+ * Initialize controller module
+ * Must be called before RTOS scheduler starts
+ * @return W_SUCCESS if initialization successful
+ */
+w_status_t controller_init(void) {
+    // Create internal state queue (length = 1)
+    internal_state_queue =
+        xQueueCreate(1, sizeof(controller_input_t)); // estimator function to controller task
+    output_queue = xQueueCreate(1, sizeof(controller_output_t));
+
+    // check queue creation
+    if (NULL == internal_state_queue || NULL == output_queue) {
+        log_text("controller", "queue creation failed");
+        return W_FAILURE;
+    }
+
+    // TODO gain instance init
+
+    // return w_status_t state
+    log_text("controller", "initialization successful");
+    return W_SUCCESS;
+}
+
+/**
+ * Update controller with new state data - called by state estimation module
+ * @param new_state Latest state estimate from state estimation
+ * @return W_FAILURE if validation/queueing fails
+ */
+w_status_t controller_update_inputs(controller_input_t *new_state) {
+    xQueueOverwrite(internal_state_queue, new_state); // overwrite internal queue
+
+    return W_SUCCESS;
+}
+
+/**
+ * Get most recent control output (cammanded angle and timestamp) - called by
+ * state estimation module
+ * @param output Pointer to store output -> type defined in controller
+ * @return W_FAILURE if no output available
+ */
+w_status_t controller_get_latest_output(controller_output_t *output) {
+    if (pdPASS == xQueuePeek(output_queue, output, pdMS_TO_TICKS(CONTROLLER_CYCLE_TIMEOUT_MS))) {
```

Comment:
`CONTROLLER_CYCLE_TIMEOUT_MS` is called that cause it defines the cycle delay of the controller (i.e. the part in the task). Even if this timeout happens to be 5 ms, you should use a different macro because they represent unrelated delays

That being said, in this case the timeout should be 0, because the queue should always be populated. You should add a call to overwrite the output queue in `controller_init`, with just "0" or smth, otherwise you will be stuck in deadlock as soon as you initialize (controller won't run until it gets input from estimator, and estimator won't run until it gets input from controller)